### PR TITLE
cuda: fix check for GPU device availability

### PR DESCRIPTION
**Pull request by rst0git**
The check for `/dev/nvidiactl` to determine if the CUDA plugin can be used is unreliable because in some cases the default path for driver installation is different [^1]. This pull request changes the logic to check if a GPU device is available in `/proc/driver/nvidia/gpus/`. This approach is similar to [`torch.cuda.is_available()`](https://pytorch.org/docs/stable/_modules/torch/cuda.html#is_available) and it is a more accurate indicator. The subsequent check for support of the `cuda-checkpoint --action` option would confirm if the driver supports checkpoint/restore.

[^1]: https://github.com/NVIDIA/gpu-operator

Fixes: #2509

### CODE REVIEW

**Review by jesus-ramos — APPROVED** [review]

File: plugins/cuda/cuda_plugin.c [hunk 1]
```diff
@@ -470,6 +471,18 @@ int cuda_plugin_resume_devices_late(int pid)
 }
 CR_PLUGIN_REGISTER_HOOK(CR_PLUGIN_HOOK__RESUME_DEVICES_LATE, cuda_plugin_resume_devices_late)
 
+/**
+ * The cuda-checkpoint tool requires the libcuda library.
```
Comment (avagin) [inline plugins/cuda/cuda_plugin.c:475]: * does it really require it? If I remember right, cude-checkpoint initially was a static-linked binary. Did I miss something?

* The idea was to find out whether a host has a required hardware. For example, user can use one vm image for all their vm-s, but only some of them will have nvidia cards. This check will always pass even when a host doesn't have any nvidia cards.

Comment (rst0git) [reply plugins/cuda/cuda_plugin.c:475]: >does it really require it?

yes, the `cuda-checkpoint` tool is linked against `libcuda.so.1` and it fails if this library is not installed.

> This check will always pass even when a host doesn't have any nvidia cards.

We can also check the content of `/proc/driver/nvidia/gpus/*`:

```
# System without GPU
$ ls /proc/driver/nvidia/gpus/*
ls: cannot access '/proc/driver/nvidia/gpus/*': No such file or directory

# System with GPU
$ ls /proc/driver/nvidia/gpus/
0000:01:00.0

$ ls  /proc/driver/nvidia/gpus/
0000:82:00.0
```

Comment (rst0git) [reply plugins/cuda/cuda_plugin.c]: I've updated the pull request to check only for `/proc/driver/nvidia/gpus/*`.

File: plugins/cuda/cuda_plugin.c [hunk 2]
```diff
@@ -470,6 +470,22 @@ int cuda_plugin_resume_devices_late(int pid)
 }
 CR_PLUGIN_REGISTER_HOOK(CR_PLUGIN_HOOK__RESUME_DEVICES_LATE, cuda_plugin_resume_devices_late)
 
+static bool has_gpu()
+{
+	DIR *dir = opendir("/proc/driver/nvidia/gpus/");
```
Comment (avagin) [inline plugins/cuda/cuda_plugin.c:475]: nit: I think it is enough to check an existence of  /proc/driver/nvidia/gpus. I don't think that someone loads a module without having any nvidia card on the host.

File: plugins/cuda/cuda_plugin.c [hunk 3]
```diff
@@ -481,10 +497,12 @@ int cuda_plugin_init(int stage)
 		}
 	}
 
-	if (!fault_injected(FI_PLUGIN_CUDA_FORCE_ENABLE) && access("/dev/nvidiactl", F_OK)) {
-		pr_info("/dev/nvidiactl doesn't exist. The CUDA plugin is disabled.\n");
-		plugin_disabled = true;
-		return 0;
+	if (!fault_injected(FI_PLUGIN_CUDA_FORCE_ENABLE)) {
```
Comment (avagin) [inline plugins/cuda/cuda_plugin.c:500]: nit: !fault_injected() && !has_gpu()

**Comment by avagin** [comment]
LGTM. Thanks.

File: plugins/cuda/cuda_plugin.c (hunk 3 above)
Comment (rst0git) [reply plugins/cuda/cuda_plugin.c:500]: Thanks, I also noticed that (if I understand correctly), setting `FI_PLUGIN_CUDA_FORCE_ENABLE` should skip the GPU device check. In other words, the condition should be:
```
!fault_injected(FI_PLUGIN_CUDA_FORCE_ENABLE) || !has_nvidia_gpu()
```
I've also updated the zdtm test to run with `--criu-plugin cuda` to test this change.

**Pull request merged**